### PR TITLE
Enhanced runner v2

### PR DESCRIPTION
**Pull request by samirans89**

### CODE REVIEW

File: src/main/java/com/browserstack/BatchExecutionRunner.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import io.cucumber.core.exception.UnrecoverableExceptions;
+import io.cucumber.core.options.RuntimeOptions;
+import io.cucumber.core.runtime.CucumberExecutionContext;
+import org.openqa.selenium.WebDriver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.*;
```
Comment (samirans89) [inline src/main/java/com/browserstack/BatchExecutionRunner.java]: let's import specific classes instead of *

File: src/main/java/com/browserstack/BatchExecutionRunner.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+            // WebDriver Creation
+            WebDriver webDriver = webDriverFactory.createWebDriverForPlatform(execution.getPlatform(), execution.getPickle().getName());
+            eventBus.send(new WebDriverCreated(Instant.now(), webDriver));
+
+            // Pickle Execution
+            runner.runPickle(execution.getPickle());
+
+            // Completed Execution
+            //LOGGER.debug("Completed {} on {}", execution.getPickle().getName(), e.getName());
```
Comment (samirans89) [inline src/main/java/com/browserstack/BatchExecutionRunner.java]: please uncomment OR remove, as applicable.

File: src/main/java/com/browserstack/BatchExecutionRunner.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        }else {
+            LOGGER.debug("Cucumber feature execution completed");
+        }
+        eventBus.send(new BuildCompleted(Instant.now()));
+    }
+
+    private void execute(Runnable runnable) {
+        try {
+            runnable.run();
+        } catch (Throwable var3) {
```
Comment (samirans89) [inline src/main/java/com/browserstack/BatchExecutionRunner.java]: let's either use default suggested name for object OR a meaningful content-based name for the object instead of `var3`

File: src/main/java/com/browserstack/WebDriverRuntime.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+import io.cucumber.core.feature.FeatureParser;
+import io.cucumber.core.filter.Filters;
+import io.cucumber.core.gherkin.Feature;
+import io.cucumber.core.gherkin.Pickle;
+import io.cucumber.core.options.RuntimeOptions;
+import io.cucumber.core.order.PickleOrder;
+import io.cucumber.core.plugin.PluginFactory;
+import io.cucumber.core.plugin.Plugins;
+import io.cucumber.core.resource.ClassLoaders;
+import io.cucumber.core.runtime.*;
```
Comment (samirans89) [inline src/main/java/com/browserstack/WebDriverRuntime.java]: change to specific classes instead of *

File: src/main/java/com/browserstack/WebDriverRuntime.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+import io.cucumber.plugin.Plugin;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.util.*;
```
Comment (samirans89) [inline src/main/java/com/browserstack/WebDriverRuntime.java]: change to specific classes instead of *

File: src/main/java/com/browserstack/WebDriverRuntime.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        CucumberExecutionContext cucumberExecutionContext = this.context;
+        Objects.requireNonNull(cucumberExecutionContext);
+        features.forEach(cucumberExecutionContext::beforeFeature);
+        List<Execution> executions = new ArrayList<>();
+        features.stream()
+                .flatMap((feature) -> {
+                    List<Pickle> pickles = feature.getPickles().stream()
+                            .filter(this.filter)
+                            .collect(Collectors.collectingAndThen(Collectors.toList(), (list) -> this.pickleOrder.orderPickles(list)))
+                            .stream().limit(this.limit > 0 ? (long) this.limit : 2147483647L)
```
Comment (samirans89) [inline src/main/java/com/browserstack/WebDriverRuntime.java]: Long.MAX_VALUE or Integer.MAX_VALUE works?

File: src/main/java/com/browserstack/WebDriverRuntime.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        // Pickle execution
+        batchExecutionRunner.submitExecutions(0, executions);
+    }
+
+
+    private void execute(Runnable runnable) {
+        try {
+            runnable.run();
+        } catch (Throwable var3) {
+            UnrecoverableExceptions.rethrowIfUnrecoverable(var3);
```
Comment (samirans89) [inline src/main/java/com/browserstack/WebDriverRuntime.java]: let's either use default suggested name for object OR a meaningful content-based name for the object instead of var3

File: src/main/java/com/browserstack/junit/NoObjectFactorya.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package com.browserstack.junit;
+
+import io.cucumber.core.backend.ObjectFactory;
+
+/**
+ * This object factory does nothing. It is solely needed for marking purposes.
```
Comment (samirans89) [inline src/main/java/com/browserstack/junit/NoObjectFactorya.java]: Is this file really needed?
any alternatives?

File: src/main/java/com/browserstack/report/CustomReportListener.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
+package com.browserstack.report;
+
+import com.browserstack.*;
```
Comment (samirans89) [inline src/main/java/com/browserstack/report/CustomReportListener.java]: replace * with actual ClassNames in all the relevant imports here

File: src/main/java/com/browserstack/report/builder/CustomReportBuilder.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        for (Embedding embedding : imageEmbeddings) {
+            writer.write("document.getElementById('");
+            writer.write(embedding.getEmbeddingId());
+            writer.write("').src='data:image;base64,");
+       /*     // Todo:
+            byte[] data = embedding.getData();
+            StringBuffer stringBuffer = new StringBuffer();
+            for (byte datum : data) {
+                stringBuffer.append(datum);
+            }*/
```
Comment (samirans89) [inline src/main/java/com/browserstack/report/builder/CustomReportBuilder.java]: please cleanup as appropriate.

File: src/main/java/com/browserstack/report/builder/HtmlReportBuilder.java [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+import com.browserstack.report.models.*;
+import com.github.mustachejava.DefaultMustacheFactory;
+import com.github.mustachejava.Mustache;
+import io.cucumber.plugin.event.HookType;
+import org.apache.commons.lang3.EnumUtils;
+
+import java.io.*;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
```
Comment (samirans89) [inline src/main/java/com/browserstack/report/builder/HtmlReportBuilder.java]: replace * with actual class names required.

File: src/main/java/com/browserstack/report/builder/HtmlReportBuilder.java [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+//            String[] keyValue = value.trim().split("=");
+//
+//            if (keyValue.length == 2) {
+//                envData.add(keyValue[0].trim() + " = " + keyValue[1].trim());
+//            }
+//        }
+
+//        if (envData.isEmpty()) {
+        envData.add("No additional environment information provided.");
+        //      }
```
Comment (samirans89) [inline src/main/java/com/browserstack/report/builder/HtmlReportBuilder.java]: Please cleanup as applicable.

**Comment by sanketsmali** [comment]
Reviewed the changes

**Review by sanketsmali — APPROVED** [review]
